### ISSUE TITLE
Why removing ServiceLocator dependency within ControllerAbstract ?

### ISSUE THREAD

**Issue by sophpie** [zendframework/zend-mvc#137]
As one of the most valuable thing of a service locator is to build service on demand, what is the purpose of removing dependency to ServiceLocator into Controllers ?

Now we have to invoke services and inject them into the controller even if at the end those services will not be called by controller.

By design, afaik, a service locator is made to be a dependency to any of its client, if not it become quite useless.

So, is there any documentation that explains why we decided to remove service locator dependency from controller ? So far, I have no arguments for my customer,and the less I can tell is I complitly  agree with them when they say the 2.5 version was better designed.

**Comment by Ocramius** [comment]
Hey @sophpie,

> what is the purpose of removing dependency to ServiceLocator into Controllers ?

The main goal is reducing run-time service-location operations. Service location causes coupling with an interface that solves a problem that is orthogonal to the currently solved problem. This means that any operation now involves coupling with:
1. a string (the name of the service) (no guarantees here)
2. a SL interface
3. the exceptions that the service locator may throw
4. the service that you actually want

All this without type-safety, whereas you really just want the 4th item.

> Now we have to invoke services and inject them into the controller even if at the end those services will not be called by controller.

That is correct, but is also a minimal problem. Huge services that get injected and that are performance-sensitive (especially at instantiation time) can be made [lazy](https://ocramius.github.io/blog/zf2-and-symfony-service-proxies-with-doctrine-proxies/) (also see http://zendframework.github.io/zend-servicemanager/lazy-services/ ), eventually. In this way, their delayed instantiation becomes a detail/optimization, rather than part of the design of our code.

> By design, afaik, a service locator is made to be a dependency to any of its client, if not it become quite useless.

Yes, but only some clients are good candidates as consumers of service-locators. Specifically, we'd talk about clients that map from something to a service via a service-locator. Good candidates are:
- the view helper system
- a dynamic class loader
- a routing system
- configurable strategy pattern implementations in general
  
  > So, is there any documentation that explains why we decided to remove service locator dependency from controller ?

I think not, but this is a good call to document why consuming the `ServiceLocator` is a bad practice in most scenarios.

> So far, I have no arguments for my customer, and the less I can tell is I completely agree with them when they say the 2.5 version was better designed.

Then my wild guess is that:
- the customer is coupling too much logic with the framework
- the customer is not testing any of his logic
- the customer didn't yet upgrade any modules providing service definitions (those break, _a lot_)

There are tons of articles on [why service location is an anti-pattern](https://www.google.com/search?q=service+location+is+an+anti-pattern&ie=utf-8&oe=utf-8), and I've seen those come true over the years in which I've been using ZF2, either as a module developer, as a consumer, as a consultant for a client using ZF2 or as a developer of the framework. Things crash and burn everywhere, and often when instantiation time is delayed to execution time.

You can bring these two pieces of code to a customer, and ask following questions:
- when can this code break?
- how many failures are possible?
- how do you make this code work? (consider yourself seeing this code for the first time)
- how do you write a test for this code?
- how do you port this code out of a zf2 context (upgrade or switch to different framework)?

``` php
class ListUsers
{
    private $sl;
    public function setServiceLocator(ContainerInterface $sl)
    {
        $this->sl = $sl;
    }
    public function __invoke()
    {
        return new ViewModel(['users' => $this->sl->get('userRepository')->findAll());
    }
}
```

``` php
class ListUsers
{
    private $userRepository;
    public function __construct(UserRepository $userRepository)
    {
        $this->userRepository = $userRepository;
    }
    public function __invoke()
    {
        return new ViewModel(['users' => $this->userRepository->findAll());
    }
}
```

I think the advantages will be clear/simple/visible to everyone, after evaluating the questions above.

**Comment by weierophinney** [comment]
@Ocramius Thanks for the write-up! I was in the middle of a lengthy one when you posted, so there's some duplication, but I'm going to post anyways.

The intent of zend-servicemanager is for use as an [Inversion of Control](https://en.wikipedia.org/wiki/Inversion_of_control) container.

It was never intended as a general purpose [service locator](https://en.wikipedia.org/wiki/Service_locator_pattern) (interestingly, that link details mostly disadvantages to the pattern!); that role was something foisted onto it in the spirit of "rapid application development" and to "simplify initial development," but the intention even there was that, once a class has stabilized, you should refactor to inject dependencies.

Why?

[Google for "service locator anti pattern"](https://www.google.com/search?q=service%20locator%20anti%20pattern) to get an idea of why it shouldn't be used. The main points boil down to:
- Dependency hiding.
- Error indirection.
- Type safety.
- Brittleness.

Let's look at each of these individually.

### Dependency hiding

What is meant by "dependency hiding?"

Take a look at this class signature:

``` php
class Foo
{
    public function dispatch(Request $request, Response $response);

    public function setServiceLocator(ServiceLocatorInterface $serviceLocator);
}
```

Based on that, you'd expect:
- that you can instantiate the object with no dependencies.
- if you feel the need to, you _could_ pass a service locator to the instance.
- you should be able to execute `dispatch()` by passing it a request and response instance, and it should successfully return.

The service locator is nebulous; its purpose isn't clear, and it's clearly not a _required_ dependency, as it's in a setter method.

So, you go and write a test for the `dispatch()` method, and you get a `ServiceNotFoundException`. What's wrong?

You dive into the code of the `dispatch()` method:

``` php
public function dispatch(Request $request, Response $response)
{
    $authentication = $this->serviceLocator->get('authentication');

    if (! $authentication->hasIdentity()) {
        $response->setStatus(401);
        return $response;
    }

    $identity = $authentication->getIdentity();
    $response->setBody(
        $this->serviceLocator->get('renderer')->render(
            'foo',
            ['identity' => $identity]
        )
    );
    return $response;
}
```

There's two possible places that `ServiceNotFoundException` may have been thrown: on the first line of the method, or within the `setBody()` call. In both cases, you're faced with a conundrum:
- You now know that the service locator is _required_. That wasn't obvious from looking at the class originally; it appeared to be an _optional_ dependency.
- You have no idea what instance types are expected for each of the `authentication` and `renderer` services.

This latter is particularly troubling. You now have to understand all the various locations within the application that services may be defined, and start hunting through those. Chances are, you'll discover that those service names may actually be _aliases_, which means you'll determine what it aliases to, but then have to re-start your search to determine what the actual service is.

That, in a nutshell, is dependency hiding. The _requirements_ for operation of the class are hidden _within_ the code, and the types cannot necessarily be inferred.

(Sure, you could add annotations above the `get()` calls to detail the types. But that's a band-aid; you still need to look in the code itself to determine what the requirements are.)

One side effect of dependency hiding is that _it makes testing more difficult_.  I think the example above illustrates that; you cannot look at the signatures alone to get an idea of behavior and requirements, but need to dig into the code. Additionally, test setup becomes more difficult, and more brittle, as you're now required to add a dependency on a container, populate the container, and hope you didn't miss something. I'll speak more on this later. The point is: anything that makes testing more difficult means developers will avoid testing, and that decreases code quality.

Defining the requirements _as constructor arguments_ makes them explicit, and ensures somebody looking solely at the API understands what is required for operation.

**tl;dr**: You want the dependency you're consuming, not the three steps of indirection required to get at it (configuring the container, injecting the container into your instance, and pulling the dependency from the container).  Make all dependencies required, and inject them in the constructor.

### Error indirection

Re-using the example from above, let's examine the fact that we got a `ServiceNotFoundException`. This is happening _at runtime_. Essentially, the work of bootstrapping, routing, instantiating the controller, and pre-dispatch listeners have already run, only to fail once we get to the actual logic requested _because a dependency was missing_.

In a typical PHP application workflow, this is not much different from if the dependencies were directly injected. But if you consider usage in a system such as [React](http://reactphp.com), where bootstrapping can occur once, it's quite problematic; it's in essence a runtime exception, _due to misconfiguration_. This is quite difficult to trace, and not something you want to have happen in production.

### Type safety

Again, going back to the original example: we don't know what the types expected are, _nor can we guarantee that what we pull from the container will be correct._

An inexperienced developer, or one not familiar with all the use cases for a given instance in a container could define a factory returning an unexpected instance. You won't know until runtime, in production, that this has occurred, when you suddenly get "method does not exist" fatal PHP errors. These are difficult to track down, as you will not know what the type is, what was expected, nor where the instance is originally defined.

Compare this to a dependency declared in the constructor:

``` php
class Foo
{
    public function __construct(AuthenticationService $authentication);
}
```

You'll still get a fatal error, but you'll know that the class was being instantiated with an invalid argument from the beginning. This type of problem can often be found with static analysis tools, giving another way for you to help improve your code quality easily.

Another aspect of this is that your IDEs will now also be able to assist you in understanding what methods are available. Because the property is injected in the constructor, the _static analysis_ (I'm using that word again!) built-in to most IDEs will be able to infer the type when you access it in your code, and give you type hinting. This is not universally true with service locators (I understand PHPStorm is making some headway on this, but I also know it's a very difficult task to accomplish, and error prone).

### Brittleness

Relying on a service locator introduces brittleness into your designs.

Every time you add a call to `get()`, you're introducing a new dependency. This often breaks tests:
- If you're mocking the service locator, you now have additional calls to its methods that may be called in the course of a test, making the mock fail its assertions.
- If you're using a concrete locator instance, and an instance is expected to be present, you now get exceptions raised over the course of test execution.

This sort of brittleness leads to developers not wanting to test, making the code more brittle and more likely to break in unexpected ways in the future.

Additionally, it leads to _undocumented requirements_, making it less clear for a consumer to know what services need to be present for the code to work. When you work across teams, this is critical.

Another aspect of using a service locator is that it's very easy for your class to grow to span too many responsibilities.

One argument often used in favor of using a service locator is _optional_ dependencies: dependencies that are only used during paths of code execution. If the dependency is particularly heavy (web services, database access, etc.), the argument is that it makes sense to pull these from the container only if they're about to be used.

There are two ways to address this:
- We already offer [lazy services](http://zendframework.github.io/zend-servicemanager/lazy-services/), which solve the problem by creating a proxy class that wraps the factory for retrieving the service. You interact with it just as you would the original instance, but the "heavy instantiation" is delayed until first use.
- Split your concerns into multiple classes! This is the better solution anyways; if you know certain dependencies are only in certain code paths, create a new controller for that path, and route specifically to it. As an example, if you know that database access will only occur on (a) POST requests to the service, and (b) when validation occurs, then:
  - Create a controller that maps specifically to POST requests for the given path, and
  - optionally, wrap the database connection as a lazy service. Chances are, though, that if you get the request routed to that specific controller, having the database access ready will be acceptable performance-wise.

**tl;dr**: dependence on service locators leads to brittle design and scope creep. When you pay attention to dependencies, you end up splitting concerns into multiple classes, making them easier to test and maintain.

### Takeaways

We introduced the `ServiceLocatorAwareInterface` due to pressure from users who wanted "rapid application development" features, and who were unclear about the benefits of dependency injection. I, for one, have regretted adding it from the time of our first v2 stable release, as I think it promotes bad habits and reduces code quality for those who rely on the pattern.

As @Ocramius has pointed out, there _are_ valid times to use a service locator.  Typically these fall into very specific categories:
- plugins/helpers/etc.
- strategies

But as a general-purpose way of getting dependencies, it's an anti-pattern at best, and leads to quality decline and brittle architecture.
